### ISSUE TITLE
[Vega-Expression] Allow accessing a signal / function only if it exists

### ISSUE THREAD

**Issue by kanitw** [vega/vega#2446]
Vega allows users to register an expression function. 

Imagine a case where users want to make a spec that flexibly works no matter the function exists or not, we need a way to call it only when it exists.

For example, if we want to use a potentially registered function `sum`, we can't do `foo ? foo(1,2) : 1+2` because we will get `unrecognized signal name: 'foo'`. 

It would be great if we could do `globals['foo'] ? globals['foo'](1,2) : 1+2`.

**Comment by kanitw** [comment]
For context, I'm planning to use this when one registers a custom format function (e.g., a wrapper to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RelativeTimeFormat), and only uses the formatter when it exists.

**Comment by domoritz** [comment]
`globals` is like `window` then, right?

**Comment by kanitw** [comment]
Yeah.

**Comment by roying** [comment]
Vega expressions (v5.10.0) include functions for testing data values, e.g.
- isDefined(value)
- isObject(value)
- isValid(value)

https://vega.github.io/vega/docs/expressions/

How about adding vega expression function(s) such as _isFunction()_ or _typeOf()_ ? For example:
```
"exp": "isFunction('myExpFunc') ? myExpFunc(1, 2) : 1 + 2"

"exp": "typeOf('myExpFunc') == 'function' ? myExpFunc(1, 2) : 1 + 2"
```

**Comment by jheer** [comment]
Hmm. Function resolution is done at parse time (e.g., functions may resolve to actual function calls, or to inlined generated code), so with the current setup we can't (in general) do this with an approach that evaluates the presence of function definitions at _run time_. That said, we could consider a special function type that acts like a compiler pre-processor at _parse time_. For example:

```
ifFunction('myExpFunc', myExpFunc(1, 2), 1 + 2)
```

In this case, the expression parser might check if the function `myExpFunc` is registered. If so, it will then parse (and codegen) the expression `myExpFunc(1, 2)`, otherwise it will default to `1 + 2`. Would something like this suffice?

If so, I believe all the necessary edits would be limited to the vega-expression package, and require a minor version bump. We should also consider what a good name for this special operation would be... I'm not very satisfied with `ifFunction`.

**Comment by kanitw** [comment]
For now, I'll make Vega-Lite only output certain custom formatType only if the formatType is also registered with Vega-Lite.  

https://github.com/vega/vega-lite/pull/6055/files#diff-6a66b8be2cdddd8881db273ef61ff401R77
This means that users need to register custom format function with both Vega View and Vega-Lite.

**Comment by mathiastiberghien** [comment]
I didn't know about this, very interesting!
I have a suggestion to be able to extend the expression functions keeping compatibility with vega editor.
Using the config property we should be able to configure the `extensibility` channel containing `functions` channel which is an array containing the functions definition.
A function should have that schema:
```
{
 "name":"function_name",
 "args":"arg1,arg2,arg3",
 "body":"return (arg1+arg2)*arg3"
}
```
This could be a sample of a valid spec:
```
{
  "$schema": "https://vega.github.io/schema/vega/v5.json",
  "width":400,
  "height":400,
  "config":
  {
    "extensibility":{
        "functions":[
            {
                "name":"custom",
                "args":"x,y",
                "body":"let z=5;return z*(x+y)"
            }
        ]
    }
},
  "signals":[
    {
      "name":"test",
      "init":"custom(2,6)"
    }
  ],
  "marks":[
    {
      "type":"text",
      "encode":{
        "enter":{
          "text":{"signal":"'40 is expected: '+test"}
        }
      }
    }
  ]
}
```
The parser should then be improved to be handle to register the functions
using something like this:
```
if (spec.config && spec.config.extensibility && spec.config.extensibility.functions && spec.config.extensibility.functions.length) {
            spec.config.extensibility.functions.forEach(f => {
                let func = new Function(f.args, f.body);
                vega.expressionFunction(f.name, func);
            });
        }
```

Of course, there are limitations with the use of Function but it should help in a lot of cases

**Comment by domoritz** [comment]
Note that internal functions can use loops and other unsafe functions, while expressions cannot. I would only want to support safe functions in a spec to prevent bad actors from adding unsafe functionality.

**Comment by mathiastiberghien** [comment]
I'm not sure to understand since a bad actor can use expressionFunctions to add unsafe functionality (the definition is outside the spec, but the call to the function is defined in the spec).
The loop was also introduced with the use of timer event as documented [here].(https://vega.github.io/vega/examples/hypothetical-outcome-plots/)

**Comment by domoritz** [comment]
Think of an application like Wikipedia where anyone can modify a spec. We need to make sure that there is no code in the spec that sends cookies or other session keys to an untrusted website or crashes the browser.

**Comment by mathiastiberghien** [comment]
I get it.
Maybe the body of the function should be not be written in java but in expression language and the parser will transform it internally has a Java function, validating the body as an expression (considering the args as valid references).
My sample becoming:
```
{
  "$schema": "https://vega.github.io/schema/vega/v5.json",
  "width":400,
  "height":400,
  "config":
  {
    "extensibility":{
        "functions":[
            {
                "name":"custom",
                "args":"x,y",
                "body":"5*(x+y)"
            }
        ]
    }
},
  "signals":[
    {
      "name":"test",
      "init":"custom(2,6)"
    }
  ],
  "marks":[
    {
      "type":"text",
      "encode":{
        "enter":{
          "text":{"signal":"'40 is expected: '+test"}
        }
      }
    }
  ]
}
```
This will allow the users to extend the expression function allowing factorization of repetitive calculations and without opening any unsafe code injection

**Comment by domoritz** [comment]
Yep, but at this point the benefit is much smaller since all you get is a shorthand. The biggest value of custom expressions for me is to be able to extend the functionality of Vega with functions that are otherwise not possible. I’m not saying there is no benefit, it’s just less.

**Comment by mathiastiberghien** [comment]
Maybe then that we should enable the extensibility dynamic parsing only
trough the api viaa method like

 vega.enableExtensibilityParsing();

The vega designer will allow it. Some user might allow it in their specific
applications, and by default it is disabled. An 'injection' spec won't work
in Wikipedia then

Le jeu. 26 mars 2020 à 17:32, Dominik Moritz <notifications@github.com> a
écrit :

**Comment by hydrosquall** [comment]
I like the idea of letting people easily supply their own custom JS functions in the spec by writing regular JS, but as Dominik notes we'd have to find some way to sandbox their eval so access to certain APIs (like network and cookies) are restricted. 

Besides use of bad APIs, there's also the idea of protecting people from performance hazards like entering an infinite loop. Maybe there's a middle ground where you can write custom functions, where the definition is made entirely of the already restricted vega-expressions syntax. 

To avoid the risk of recursion problems, we would have to avoid letting you reference other custom inlined functions, but it would improve the legibility of expressions that get used repeatedly.
